### PR TITLE
feat(user-state): load and change state, client timer

### PR DESCRIPTION
**Pull request by mkesavan13**
# Closes [SPARK-584812](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-584812)

# This PR addresses

- User State KS UI
- User State change functionality
- Show User State Widget in the Sample app using the `onLogin` callback of Station Login Widget
- Elapsed time of currently set state.
- Respective Unit Tests

# Pending Items

- KeepAlive for Timer (Working on understanding WxCC Code Changes)

# Manual Tests Done

- Changing the Agent State
- Adding a custom agent state on the admin portal, listing it and changing to that state from the Widget
- Changing the Agent State via the Widget and Seeing it reflected in the WxCC Desktop

## Release Notes

- **New Features**
	- Introduced user login state management, allowing conditional rendering of user-related components.
	- Enhanced `UserState` component to utilize specific properties for better data handling.
	- Added new properties for agent status management, including elapsed time and error messages.

- **Bug Fixes**
	- Improved error handling in the login/logout processes to prevent runtime issues.

- **Tests**
	- Expanded test coverage for the `useUserState` hook and `UserState` component, ensuring robust functionality and error handling.

### CODE REVIEW

**Comment by mkesavan13** [comment]
@coderabbitai review

File: docs/web-component-samples/app.js [hunk 1]
```diff
@@ -20,16 +20,18 @@ function initWidgets(){
...
-        widgetsContainer.classList.remove('disabled');
+        ccStationLogin.classList.remove('disabled');
     }).catch((error) => {
         console.error('Failed to initialize widgets:', error);
     });
 }
 
 function loginSuccess(){
     console.log('Agent login has been succesful');
+    ccUserState.classList.remove('disabled');
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply docs/web-component-samples/app.js:31]: This is a sample app change. Not going to do it.

File: docs/web-component-samples/app.js [hunk 2]
```diff
@@ -1,6 +1,6 @@
-const widgetsContainer = document.getElementById('widgets-container');
 const accessTokenElem = document.getElementById('access_token_elem');
 const ccStationLogin = document.getElementById('cc-station-login');
+const ccUserState = document.getElementById('cc-user-state');
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply docs/web-component-samples/app.js:3]: Addressed this with a console error and here's the code,

```js
if (!ccStationLogin && !ccUserState) {
    console.error('Failed to find the required elements');
}
```

File: docs/react-samples/src/App.tsx [hunk 3]
```diff
@@ -16,10 +17,12 @@ function App() {
 
   const onLogin = () => {
     console.log('Agent login has been succesful');
+    setIsLoggedIn(true);
   }
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply docs/react-samples/src/App.tsx:21]: Again, not going to do this as it is only in sample app

File: packages/contact-center/user-state/src/helper.ts [hunk 4]
```diff
@@ -1,11 +1,44 @@
...
+  useEffect(() => {
+    // Reset the timer whenever the component mounts or the state changes
+    setElapsedTime(0);
+    const timer = setInterval(() => {
+      setElapsedTime(prevTime => prevTime + 1);
+    }, 1000);
+
+    // Cleanup the timer on component unmount
+    return () => clearInterval(timer);
+  }, []);
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/user-state/src/helper.ts:19]: I guess the reset is taken care in the state change success. So, this change isn't required

File: packages/contact-center/user-state/tests/user-state/index.tsx [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+describe('UserState Component', () => {
+  it('renders UserStatePresentational with correct props', () => {
+    const useUserStateSpy = jest.spyOn(helper, 'useUserState');
+    
+    render(<UserState/>);
+
+    expect(useUserStateSpy).toHaveBeenCalledWith({cc: {}});
+    const heading = screen.getByTestId('user-state-title');
+    expect(heading).toHaveTextContent('Agent State');
+  });
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/user-state/tests/user-state/index.tsx]: Addressed

File: packages/contact-center/user-state/tests/helper.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import { renderHook } from '@testing-library/react-hooks';
+import { act } from 'react'; // Import act from react
+import { useUserState } from '../src/helper'; // adjust the path accordingly
+
+jest.useFakeTimers();
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/user-state/tests/helper.ts]: Addressed and thanks for this suggestion. I've been sitting on solving this for at least an hour and a half yesterday

File: packages/contact-center/user-state/src/user-state/user-state.presentational.tsx [hunk 7]
```diff
@@ -1,14 +1,126 @@
...
+      color: isSettingAgentStatus ? 'grey' : 'black'
+    } as CSSProperties
+  };
+
+  const formatTime = (time) => {
+    const hours = Math.floor(time / 3600);
+    const minutes = Math.floor((time % 3600) / 60);
+    const seconds = time % 60;
+    return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
+  };
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/user-state/src/user-state/user-state.presentational.tsx]: Addressed

File: packages/contact-center/user-state/src/user-state/user-state.presentational.tsx [hunk 8]
```diff
@@ -1,14 +1,126 @@
...
+                  <option
+                    key={index}
+                    value={code.id}
+                  >
+                      {code.name}
+                  </option>
+                ): (
+                  <></>
+                );
+              })}
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/user-state/src/user-state/user-state.presentational.tsx]: Addressed

File: packages/contact-center/user-state/src/user-state/user-state.presentational.tsx [hunk 9]
```diff
@@ -1,14 +1,126 @@
...
+      borderRadius: '4px'
+    },
+
+    elapsedTime: {
+      position: 'absolute',
+      right: '30px',
+      top: '25px',
+      color: isSettingAgentStatus ? 'grey' : 'black'
+    } as CSSProperties
+  };
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/user-state/src/user-state/user-state.presentational.tsx]: Addressed

File: packages/contact-center/station-login/tests/helper.ts [hunk 10]
```diff
@@ -163,4 +185,21 @@ describe('useStationLogin Hook', () => {
...
+    );
+
+    act(() => {
+      result.current.logout();
+    });
+
+    await waitForNextUpdate();
+
+    expect(logoutCb).not.toHaveBeenCalled();
+  });
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/station-login/tests/helper.ts]: Addressed

File: packages/contact-center/station-login/tests/helper.ts [hunk 11]
```diff
@@ -79,6 +83,24 @@ describe('useStationLogin Hook', () => {
...
+    );
+
+    act(() => {
+      result.current.login();
+    });
+
+    await waitForNextUpdate();
+
+    expect(loginCb).not.toHaveBeenCalled();
+  });
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply packages/contact-center/station-login/tests/helper.ts]: Addressed

**Review by sreenara — CHANGES_REQUESTED** [review]
@mkesavan13 please add the tests to the PR description that you've done as well.

**Comment by mkesavan13** [comment]
> @mkesavan13 please add the tests to the PR description that you've done as well.

Done, Sreekanth

File: packages/contact-center/user-state/src/helper.ts [hunk 12]
```diff
@@ -1,11 +1,44 @@
...
+  const [errorMessage, setErrorMessage] = useState('');
+
+  const [elapsedTime, setElapsedTime] = useState(0);
+
+  useEffect(() => {
+    // Reset the timer whenever the component mounts or the state changes
+    setElapsedTime(0);
+    const timer = setInterval(() => {
+      setElapsedTime(prevTime => prevTime + 1);
+    }, 1000);
```
Comment (mkesavan13) [inline packages/contact-center/user-state/src/helper.ts:15]: On WC, the timer starts when the component is already in the DOM making it not a 0 seconds timer when appearing. Needs to be fixed

File: packages/contact-center/user-state/src/user-state/use-state.types.ts [hunk 13]
```diff
@@ -1,19 +1,19 @@
...
 
-  /**
-   * Setter for agent state
-   */
-  setAgentStatus: () => void
+  isSettingAgentStatus: boolean;
+
+  errorMessage: string;
+
+  elapsedTime: number;
```
Comment (mkesavan13) [inline packages/contact-center/user-state/src/user-state/use-state.types.ts:18]: add description (ts doc)

File: packages/contact-center/user-state/tests/helper.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+      result.current.setAgentStatus({ auxCodeId: '001', state: 'Available' });
+    });
+
+    expect(result.current.isSettingAgentStatus).toBe(true);
+
+    waitFor(() => {
+      expect(result.current.isSettingAgentStatus).toBe(false);
+      expect(result.current.errorMessage).toBe('Error: Network error');
+    });
+  });
```
Comment (mkesavan13) [inline packages/contact-center/user-state/tests/helper.ts:72]: Add tests for error timer scenario

File: packages/contact-center/station-login/src/helper.ts [hunk 15]
```diff
@@ -17,7 +17,9 @@ export const useStationLogin = (props: UseStationLoginProps) => {
     cc.stationLogin({teamId: team, loginOption: deviceType, dialNumber: dialNumber})
       .then((res: StationLoginSuccess) => {
         setLoginSuccess(res);
-        loginCb();
+        if(loginCb){
```
Comment (Kesari3008) [inline packages/contact-center/station-login/src/helper.ts:20]: Why did we make these callbacks optional?

File: packages/contact-center/station-login/src/helper.ts [hunk 16]
```diff
@@ -28,8 +30,10 @@ export const useStationLogin = (props: UseStationLoginProps) => {
     cc.stationLogout({logoutReason: 'User requested logout'})
       .then((res: StationLogoutSuccess) => {
         setLogoutSuccess(res);
-        logoutCb();
-      }).catch((error: any) => {
+        if(logoutCb){
+          logoutCb();
+        }
+      }).catch((error: Error) => {
         console.error(error);
```
Comment (Kesari3008) [inline packages/contact-center/station-login/src/helper.ts:37]: As of now, we are using console.log and console.error but are we planning to have some some sort of logger module implemented?

File: packages/contact-center/store/src/store.ts [hunk 17]
```diff
@@ -24,6 +26,8 @@ class Store implements IStore {
     return this.cc.register().then((response: Profile) => {
       this.teams = response.teams;
```
Comment (Kesari3008) [inline packages/contact-center/store/src/store.ts:27]: Does it make sense to copy the whole profile now that we are fetching more and more data ?

File: packages/contact-center/user-state/src/helper.ts [hunk 18]
```diff
@@ -1,11 +1,56 @@
...
+    setCurrentState(selectedCode);
+    const chosenState = state === 'Available' ? 'Available' : 'Idle';
+    cc.setAgentState({state: chosenState, auxCodeId, agentId, lastStateChangeReason: state}).then((response) => {
+      setIsSettingAgentStatus(false);
+      setErrorMessage('');
+      setElapsedTime(0);
+    }).catch(error => {
+      setCurrentState(oldState);
+      setErrorMessage(error.toString());
+      setIsSettingAgentStatus(false);
```
Comment (Kesari3008) [inline packages/contact-center/user-state/src/helper.ts:43]: This is being set to false irrespective of success or failure here so we can just put it once to false and I see this is only being used for CSS purpose. Can we not use any of other parameters to control the CSS?

File: packages/contact-center/user-state/tests/helper.ts [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+
+  afterEach(() => {
+    jest.useRealTimers();
+  });
+
+  it('should initialize with default values', () => {
+    const { result } = renderHook(() => useUserState({ idleCodes, agentId, cc: mockCC }));
+
+    expect(result.current.isSettingAgentStatus).toBe(false);
```
Comment (Kesari3008) [inline packages/contact-center/user-state/tests/helper.ts:28]: Instead of checking each prop one by one, we can just add complete prop check at once

**Review by Kesari3008 — COMMENTED** [review]
Minor comments

File: packages/contact-center/user-state/src/helper.ts (hunk 18 above)
Comment (Shreyas281299) [reply packages/contact-center/user-state/src/helper.ts:43]: We can put this in .finally()

Comment (Shreyas281299) [reply packages/contact-center/user-state/src/helper.ts:43]: OPEN ITEM: In the promise we might have a way to figure out if a promise is in the pending state, which will eliminate the use of IsSettingAgentStatus

File: docs/web-component-samples/index.html [hunk 20]
```diff
@@ -22,9 +22,9 @@ <h1>Contact Center widgets as web-component</h1>
       autocapitalize="off"
     />
     <button onclick="initWidgets()" disabled>Init Widgets</button>
-    <div id="widgets-container" class="disabled">
-      <widget-cc-station-login id="cc-station-login"></widget-cc-station-login>
-      <widget-cc-user-state></widget-cc-user-state>
+    <div id="widgets-container">
+      <widget-cc-station-login class="disabled" id="cc-station-login"></widget-cc-station-login>
+      <widget-cc-user-state class="disabled" id="cc-user-state"></widget-cc-user-state>
```
[the remark this answers is not in the document]
Comment (mkesavan13) [reply docs/web-component-samples/index.html]: Done.

File: packages/contact-center/user-state/src/user-state/use-state.types.ts (hunk 13 above)
Comment (mkesavan13) [reply packages/contact-center/user-state/src/user-state/use-state.types.ts]: Done

File: packages/contact-center/user-state/tests/helper.ts (hunk 19 above)
Comment (mkesavan13) [reply packages/contact-center/user-state/tests/helper.ts:28]: Done

File: packages/contact-center/user-state/tests/user-state/index.tsx [hunk 21]
```diff
@@ -0,0 +1,24 @@
+import React from 'react';
+import {render, screen} from '@testing-library/react';
+import {UserState} from '../../src';
+import * as helper from '../../src/helper';
+import '@testing-library/jest-dom';
+
+// Mock the store import
+jest.mock('@webex/cc-store', () => {return {
+  cc: {},
+  idleCodes: [],
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/contact-center/user-state/tests/user-state/index.tsx:10]: Not required, Its coming from a mock

File: packages/contact-center/user-state/tests/helper.ts [hunk 22]
```diff
@@ -0,0 +1,90 @@
...
+
+  const agentId = 'agent123';
+
+  beforeEach(() => {
+    jest.useFakeTimers();
+    mockCC.setAgentState.mockReset();
+  });
+
+  afterEach(() => {
+    jest.useRealTimers();
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/contact-center/user-state/tests/helper.ts:22]: clearAllTimers will clear the timers not reset them

File: packages/contact-center/user-state/src/user-state/user-state.presentational.tsx [hunk 23]
```diff
@@ -1,14 +1,123 @@
...
+});
+
 const UserStatePresentational: React.FunctionComponent<IUserState> = (props) => {
-  const {handleAgentStatus, setAgentStatus} = props;
+  const {idleCodes,setAgentStatus,isSettingAgentStatus, errorMessage, elapsedTime, currentState} = props;
+
+  const styles = useMemo(() => getStyles(isSettingAgentStatus), [isSettingAgentStatus]);
+  const selectRef = useRef<HTMLSelectElement>(null);
+
+  const formatTime = (time: number): string => {
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/contact-center/user-state/src/user-state/user-state.presentational.tsx:77]: formatTime is only used in presentational. Its fine to keep it here

File: packages/contact-center/user-state/src/helper.ts [hunk 24]
```diff
@@ -1,11 +1,44 @@
...
+  }) => {
+    setIsSettingAgentStatus(true);
+    const chosenState = state === 'Available' ? 'Available' : 'Idle';
+    cc.setAgentState({state: chosenState, auxCodeId, agentId, lastStateChangeReason: state}).then((response) => {
+      setIsSettingAgentStatus(false);
+      setErrorMessage('');
+      setElapsedTime(0);
+    }).catch(error => {
+      setErrorMessage(error.toString());
+      setIsSettingAgentStatus(false);
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/contact-center/user-state/src/helper.ts:43]: Issue is fixed. The timmer is set to 0 only if the state change is successfull

File: packages/contact-center/user-state/src/helper.ts (hunk 12 above)
Comment (Shreyas281299) [reply packages/contact-center/user-state/src/helper.ts:16]: Fixed this

File: packages/contact-center/store/tests/store.ts [hunk 25]
```diff
@@ -47,14 +47,18 @@ describe('Store', () => {
     it('should set teams and loginOptions on successful register', async () => {
       const mockResponse = {
         teams: [{ id: 'team1', name: 'Team 1' }],
-        loginVoiceOptions: ['option1', 'option2']
+        loginVoiceOptions: ['option1', 'option2'],
+        idleCodes: [{ id: 'code1', name: 'Code 1', isSystem: false, isDefault: false }],
+        agentId: 'agent1'
```
[the remark this answers is not in the document]
Comment (Shreyas281299) [reply packages/contact-center/store/tests/store.ts]: Updated

File: packages/contact-center/store/src/store.ts (hunk 17 above)
Comment (Shreyas281299) [reply packages/contact-center/store/src/store.ts:27]: Copying the whole profile will cause unnecessary rerender on each widgets.

**Review by Shreyas281299 — APPROVED** [review]

**Pull request merged**